# Frequent nouns WE

In [42]:
import pandas as pd
import numpy as np

We extracted a list of most frequently used nouns and we can merge them with the WE of FlauBERT.

In [23]:
top2000n = pd.read_csv('Most Frequent Words/top1000_NOUN.csv', index_col='Word')[:2000]

In [24]:
top1000n

lemme cgram  freqlemlivres
Word                                    
homme         homme   NOM        1398.85
hommes        homme   NOM        1398.85
jour           jour   NOM        1341.76
jours          jour   NOM        1341.76
temps         temps   NOM        1289.39
...             ...   ...            ...
cesses        cesse   NOM          71.96
flamme       flamme   NOM          71.82
flammes      flamme   NOM          71.82
pantalon   pantalon   NOM          71.76
pantalons  pantalon   NOM          71.76

[1000 rows x 3 columns]

In [19]:
all_n_we = pd.read_csv('all_nouns_we.csv', index_col=0)

Out of 1000 most popular nouns, 899 have their own embedding. We can merge them into one DataFrame.

In [28]:
freq_n_we = all_n_we.merge(top1000n, left_index=True, right_index=True)
freq_n_we

0         1         2         3         4         5  \
absence     3.322297 -0.768149  3.555035 -0.685460  1.201102 -3.151831   
absences    3.295355 -0.794079  2.472845  3.223941  5.035848 -4.848855   
accord     -0.732188 -2.024965  1.674970 -4.194049 -2.528312 -2.400746   
accords     1.037610 -0.246123  3.333316 -4.094496  0.056898 -2.978131   
action      1.376720 -3.602622  0.270682 -7.004842  1.199668 -0.341319   
...              ...       ...       ...       ...       ...       ...   
événements -1.245101 -0.820061  1.330310  5.786696  3.966537 -3.242018   
être        0.166950 -1.121742  3.855680  0.846314 -3.065409  0.577782   
êtres       3.589590 -4.551544  6.088149  2.847501  3.829500 -3.342109   
île        -0.489183 -3.752668 -2.676017  4.677486  1.093927 -5.670148   
îles       -2.838814 -3.576476 -1.082639  5.224788  1.871563 -5.175120   

                   6         7         8         9  ...       507       508  \
absence    -3.592423  1.189369 -2.742928 -2.666739  ... -3.189504 -1.781391   
absences   -4.557629 -2.719433 -1.932022 -6.440993  ... -2.720232 -5.453354   
accord     -4.672920  0.374903  6.015941 -0.066068  ... -1.511964 -3.012614   
accords    -4.198946 -2.159259  3.082677  2.101236  ...  1.128479 -2.874813   
action      3.974475 -0.292309  2.597899  1.607074  ... -3.168099 -2.357484   
...              ...       ...       ...       ...  ...       ...       ...   
événements -0.265572  2.159446  0.661078 -0.936386  ... -1.882473 -2.838502   
être        1.696410  0.962429  1.131322 -5.577701  ... -4.168539 -1.392100   
êtres       2.749030 -0.169812  1.994654  0.893699  ...  0.022459 -3.056951   
île        -0.456426  2.867334  2.531248 -2.738322  ... -4.173942 -3.766897   
îles        2.896314  1.005481  4.880717 -2.816181  ... -2.302102 -5.743091   

                 509       510       511        gn     gender      lemme  \
absence    -1.852207  5.123568 -2.381779  singular   feminine    absence   
absences   -1.545623  5.919082  0.577700    plural   feminine    absence   
accord     -2.688764 -5.424189 -6.660925  singular  masculine     accord   
accords    -3.825052 -5.254343 -6.959417    plural  masculine     accord   
action     -0.797240  3.015040  0.167511  singular   feminine     action   
...              ...       ...       ...       ...        ...        ...   
événements -6.080573  1.265242  1.124227    plural  masculine  événement   
être       -5.050613  1.669239 -4.337866  singular  masculine       être   
êtres      -1.911428  0.167977 -1.285041    plural  masculine       être   
île        -7.543975 -2.131077 -4.665631  singular   feminine        île   
îles       -6.689997  0.354014 -2.508205    plural   feminine        île   

            cgram  freqlemlivres  
absence       NOM          76.28  
absences      NOM          76.28  
accord        NOM         136.15  
accords       NOM         136.15  
action        NOM          87.43  
...           ...            ...  
événements    NOM          84.59  
être          NOM         122.57  
êtres         NOM         122.57  
île           NOM         108.24  
îles          NOM         108.24  

[899 rows x 517 columns]

We can also create a dataframe with WE of all available nouns and their frequencies, having approximately 11K nouns WE available:

In [34]:
all_nouns_with_freq = pd.read_csv('Most Frequent Words/top1000_NOUN.csv', index_col='Word')
all_n_we_with_freq = all_n_we.merge(all_nouns_with_freq, left_index=True, right_index=True)
all_n_we_with_freq

0         1         2         3         4         5  \
abaisse      4.652041 -4.028067  0.883202  4.782082 -2.294611 -3.894453   
abaissement  3.209496 -2.750259  2.293690  0.207780 -1.170953 -5.999461   
abandon      3.841302 -2.121901 -1.159175 -5.437495 -1.977204 -4.404269   
abat         7.341515 -1.911543 -3.165352  0.881856  1.255612 -2.573492   
abats        2.571915 -0.997795  2.213263  7.866621  5.011806 -1.280126   
...               ...       ...       ...       ...       ...       ...   
êtres        3.589590 -4.551544  6.088149  2.847501  3.829500 -3.342109   
île         -0.489183 -3.752668 -2.676017  4.677486  1.093927 -5.670148   
îles        -2.838814 -3.576476 -1.082639  5.224788  1.871563 -5.175120   
îlot         5.139837 -7.427926  4.884292  1.064615  1.387612 -4.198579   
îlots        0.316387 -5.135355  3.555227  3.901632  0.842719 -6.265700   

                    6         7         8         9  ...       507       508  \
abaisse     -0.810279 -0.713933  4.819906  4.090149  ... -5.865204 -7.542219   
abaissement -0.837151  4.209501  0.482295  2.926001  ... -3.568309 -7.705732   
abandon     -5.100209  2.764617 -2.829459 -1.417439  ... -7.677973 -4.133019   
abat        -0.540472 -2.959341  2.970956  1.989334  ... -2.878892 -6.178219   
abats       -1.463984 -5.969271 -0.341902 -3.891024  ... -0.574585 -5.538427   
...               ...       ...       ...       ...  ...       ...       ...   
êtres        2.749030 -0.169812  1.994654  0.893699  ...  0.022459 -3.056951   
île         -0.456426  2.867334  2.531248 -2.738322  ... -4.173942 -3.766897   
îles         2.896314  1.005481  4.880717 -2.816181  ... -2.302102 -5.743091   
îlot        -1.997534  4.594993  2.311641  1.507236  ... -1.106573 -1.279163   
îlots       -2.832418  3.986920  2.747350  1.586111  ... -3.144311 -1.404029   

                  509       510       511        gn     gender        lemme  \
abaisse     -4.662949 -0.546076 -1.836027  singular   feminine      abaisse   
abaissement -5.310164  1.347664 -4.607303  singular  masculine  abaissement   
abandon     -2.451917  3.346283 -3.959826  singular  masculine      abandon   
abat        -0.410775  0.161333  3.765968  singular  masculine         abat   
abats        2.004640  1.403861 -8.578702    plural  masculine         abat   
...               ...       ...       ...       ...        ...          ...   
êtres       -1.911428  0.167977 -1.285041    plural  masculine         être   
île         -7.543975 -2.131077 -4.665631  singular   feminine          île   
îles        -6.689997  0.354014 -2.508205    plural   feminine          île   
îlot        -0.509667  1.712433 -3.325491  singular  masculine         îlot   
îlots       -1.362893 -0.495850  0.191013    plural  masculine         îlot   

             cgram  freqlemlivres  
abaisse        NOM           0.00  
abaissement    NOM           2.16  
abandon        NOM          27.36  
abat           NOM           1.01  
abats          NOM           1.01  
...            ...            ...  
êtres          NOM         122.57  
île            NOM         108.24  
îles           NOM         108.24  
îlot           NOM          10.68  
îlots          NOM          10.68  

[11395 rows x 517 columns]

# Frequency dimension

There was an assumption that dimensions 371 and 52 can be correlated to frequency. We can look at the correlations of these dimensions with `freqlemlivres`:

In [35]:
all_n_we_with_freq.loc[:, [ '52', '371', 'freqlemlivres']].corr()

52       371  freqlemlivres
52             1.000000 -0.678549      -0.101934
371           -0.678549  1.000000       0.072126
freqlemlivres -0.101934  0.072126       1.000000

The correlation between the dimension values and the frequency values is not strong. We can try to find a dimension with the strongest correlation with the values of the frequency column.

In [50]:
freq_corr_df = pd.DataFrame(columns=['corr_with_freq'])
freq_list = []

# Go through dimensions from 0 to 511
for i in range(512):
    # Find the correlation of the current column with "freqlemlivres" and add to the list
    freq_list.append(all_n_we_with_freq.loc[:, [str(i), 'freqlemlivres']].corr().iloc[0][-1])

# Use the list values as 
freq_corr_df['corr_with_freq'] = freq_list
freq_corr_df['sign'] = np.sign(freq_df['corr_with_freq'])
freq_corr_df['abs'] = abs(freq_corr_df['corr_with_freq'])
freq_corr_df

corr_with_freq  sign       abs
0         -0.056975  -1.0  0.056975
1          0.054485   1.0  0.054485
2         -0.014856  -1.0  0.014856
3         -0.053013  -1.0  0.053013
4         -0.020633  -1.0  0.020633
..              ...   ...       ...
507       -0.072089  -1.0  0.072089
508        0.013247   1.0  0.013247
509       -0.018811  -1.0  0.018811
510        0.049783   1.0  0.049783
511        0.026257   1.0  0.026257

[512 rows x 3 columns]

In [56]:
freq_corr_df.sort_values('abs').iloc[-10:]

corr_with_freq  sign       abs
208        0.106921   1.0  0.106921
469        0.108511   1.0  0.108511
113        0.108816   1.0  0.108816
91        -0.109126  -1.0  0.109126
43         0.109958   1.0  0.109958
143       -0.110892  -1.0  0.110892
314        0.125546   1.0  0.125546
324       -0.129692  -1.0  0.129692
224       -0.131444  -1.0  0.131444
146       -0.133998  -1.0  0.133998

We can see that even the dimensions with highest correlation to the frequency column are not that high, and dimension `52` is not far away from the values seen above.

This experiment doesn't allow us to establish a strong correlation with frequency for any dimension.

# Dimensions with highest and lowest avg

We can make an assumption that dimensions that have highest or lowest values can be representative of noun-ness.

For the experiment with dimensions we will take the most frequent nouns which as well *can't be interpreted as verb* (as seen above, one of the most common nouns is `être` which will potentially has dimensions both representing nouns and verbs).

For this we will first the list of nouns and verbs and find the nouns that are uniquely in the list of nouns.

In [57]:
all_verbs_with_freq = pd.read_csv('Most Frequent Words/top1000_VERB.csv', index_col='Word')
all_verbs_with_freq

lemme cgram  freqlemlivres
Word                                       
es                être   VER       15085.47
est               être   VER       15085.47
furent            être   VER       15085.47
fus               être   VER       15085.47
fusse             être   VER       15085.47
...                ...   ...            ...
éliciter      éliciter   VER           0.00
émuler          émuler   VER           0.00
émulsifier  émulsifier   VER           0.00
équeuter      équeuter   VER           0.00
équivoquer  équivoquer   VER           0.00

[64929 rows x 3 columns]

In [62]:
top1000_unique_nouns = all_nouns_with_freq[~all_nouns_with_freq.index.isin(all_verbs_with_freq.index)][:1000]
top1000_unique_nouns

lemme cgram  freqlemlivres
Word                                  
homme       homme   NOM        1398.85
hommes      homme   NOM        1398.85
jour         jour   NOM        1341.76
jours        jour   NOM        1341.76
temps       temps   NOM        1289.39
...           ...   ...            ...
actes        acte   NOM          57.30
paysan     paysan   NOM          57.30
paysanne   paysan   NOM          57.30
paysannes  paysan   NOM          57.30
paysans    paysan   NOM          57.30

[1000 rows x 3 columns]

Now we can find the WE for these nouns that can't be confused as verbs:

In [63]:
unique_freq_n_we = all_n_we.merge(top1000_unique_nouns, left_index=True, right_index=True)
unique_freq_n_we

0         1         2         3         4         5  \
absence     3.322297 -0.768149  3.555035 -0.685460  1.201102 -3.151831   
absences    3.295355 -0.794079  2.472845  3.223941  5.035848 -4.848855   
accord     -0.732188 -2.024965  1.674970 -4.194049 -2.528312 -2.400746   
accords     1.037610 -0.246123  3.333316 -4.094496  0.056898 -2.978131   
acte        5.714411 -3.455246  1.156246  0.894613 -2.072708  1.619847   
...              ...       ...       ...       ...       ...       ...   
événement  -0.693584 -0.753810  0.124109  2.200120  1.585398 -2.354075   
événements -1.245101 -0.820061  1.330310  5.786696  3.966537 -3.242018   
êtres       3.589590 -4.551544  6.088149  2.847501  3.829500 -3.342109   
île        -0.489183 -3.752668 -2.676017  4.677486  1.093927 -5.670148   
îles       -2.838814 -3.576476 -1.082639  5.224788  1.871563 -5.175120   

                   6         7         8         9  ...       507       508  \
absence    -3.592423  1.189369 -2.742928 -2.666739  ... -3.189504 -1.781391   
absences   -4.557629 -2.719433 -1.932022 -6.440993  ... -2.720232 -5.453354   
accord     -4.672920  0.374903  6.015941 -0.066068  ... -1.511964 -3.012614   
accords    -4.198946 -2.159259  3.082677  2.101236  ...  1.128479 -2.874813   
acte       -2.930497 -1.164060  1.456195 -3.298590  ... -3.608866 -1.903922   
...              ...       ...       ...       ...  ...       ...       ...   
événement  -2.994415  1.552121  4.762497 -1.361055  ... -4.292593 -4.852169   
événements -0.265572  2.159446  0.661078 -0.936386  ... -1.882473 -2.838502   
êtres       2.749030 -0.169812  1.994654  0.893699  ...  0.022459 -3.056951   
île        -0.456426  2.867334  2.531248 -2.738322  ... -4.173942 -3.766897   
îles        2.896314  1.005481  4.880717 -2.816181  ... -2.302102 -5.743091   

                 509       510       511        gn     gender      lemme  \
absence    -1.852207  5.123568 -2.381779  singular   feminine    absence   
absences   -1.545623  5.919082  0.577700    plural   feminine    absence   
accord     -2.688764 -5.424189 -6.660925  singular  masculine     accord   
accords    -3.825052 -5.254343 -6.959417    plural  masculine     accord   
acte       -1.031769  1.625798 -3.859723  singular  masculine       acte   
...              ...       ...       ...       ...        ...        ...   
événement  -5.389591  0.740749 -3.701469  singular  masculine  événement   
événements -6.080573  1.265242  1.124227    plural  masculine  événement   
êtres      -1.911428  0.167977 -1.285041    plural  masculine       être   
île        -7.543975 -2.131077 -4.665631  singular   feminine        île   
îles       -6.689997  0.354014 -2.508205    plural   feminine        île   

            cgram  freqlemlivres  
absence       NOM          76.28  
absences      NOM          76.28  
accord        NOM         136.15  
accords       NOM         136.15  
acte          NOM          57.30  
...           ...            ...  
événement     NOM          84.59  
événements    NOM          84.59  
êtres         NOM         122.57  
île           NOM         108.24  
îles          NOM         108.24  

[885 rows x 517 columns]

## Dimensions with the highest average

Now that we have a dataframe, containing words that serve as nouns only and are relatively frequent, we can analyse the dimensions. First, we can find top 5 dimensions with the highest values:

In [72]:
unique_freq_n_we.loc[:, :'511'].mean().sort_values()[-5:]

195     5.924231
326     6.414400
44      7.174508
250    10.422687
371    37.350416
dtype: float64

Some of the dimensions we have previously seen as having the highest values: `371`, `250`, `44`, `195`: [here](https://github.com/ClementineBleuze/WordEmbeddings/blob/main/Morphalou3/Investigating%20noun%20dimensions.ipynb).

Dimension `326` was not previously investiagted, we can have a look at the top 20 nouns with highest values in the dimension `326` and botton 20 nouns with the lowest values in `326`:

In [73]:
unique_freq_n_we['326'].sort_values()[-20:]

patte       12.198476
photos      12.265883
objet       12.278166
fleurs      12.279531
poil        12.297997
élément     12.353693
poils       12.598932
magasins    12.629414
fers        12.740904
chaînes     12.746919
chemise     12.968505
visage      13.011158
position    13.245049
toile       13.330914
pantalon    13.465014
chapeaux    13.678488
paysage     13.763266
cheveu      13.962662
fil         15.256657
vêtement    16.344282
Name: 326, dtype: float64

In [74]:
unique_freq_n_we['326'].sort_values()[:20]

études        -1.676383
intention     -1.597302
morceau       -1.195826
appel         -1.108422
quart         -0.810918
cri           -0.510924
étude         -0.468958
quarts        -0.342226
notes         -0.248693
anglaises     -0.035159
espoirs        0.218262
accord         0.303974
fin            0.403139
accords        0.427793
professeure    0.493299
or             0.518524
efforts        0.557973
moitié         0.578965
airs           0.658743
noirs          0.823840
Name: 326, dtype: float64

## Dimensions with the lowest average

In [75]:
unique_freq_n_we.loc[:, :'511'].mean().sort_values()[:5]

52    -10.950035
274    -8.759658
181    -6.220996
138    -5.789016
355    -5.593191
dtype: float64

Example of nouns for highest and lowest values of `52`, `274` and `138` can be found [here](https://github.com/ClementineBleuze/WordEmbeddings/blob/main/Morphalou3/Investigating%20noun%20dimensions.ipynb).

Dimensions `181` and `355` were not looked at before and we can look at them righ now.

### Dimension 181

In [76]:
unique_freq_n_we['181'].sort_values()[-20:]

forêt         -1.740409
polices       -1.695406
eaux          -1.682395
hauteur       -1.644578
détail        -1.639662
droite        -1.561079
eau           -1.356558
numéro        -1.291714
amitiés       -1.290581
courage       -1.285148
amitié        -1.267396
françaises    -1.247621
oncles        -1.206256
coin          -0.834102
kilomètre     -0.752624
boulevard     -0.737056
française     -0.638921
couvertures   -0.582790
feuilles      -0.506102
numéros       -0.388196
Name: 181, dtype: float64

In [78]:
unique_freq_n_we['181'].sort_values()[:20]

vêtement      -15.828648
pantalons     -13.349597
seins         -13.125753
chemises      -12.764800
pantalon      -12.581136
visage        -12.477160
premiers      -12.455402
avion         -12.297463
garçons       -12.292806
robes         -12.170732
instants      -12.073190
vêtements     -12.022993
nuages        -11.975925
silhouettes   -11.937696
sons          -11.860026
paupières     -11.577159
lunettes      -11.535264
tête          -11.280148
train         -11.264141
veste         -11.236717
Name: 181, dtype: float64

We can notice that all values of dimension `181` are negative and we can hypothesise that positive values of this dimension can correspond to non-nouns.

### Dimenension 355

In [79]:
unique_freq_n_we['355'].sort_values()[-20:]

amitiés         0.466341
vieilles        0.589841
maître          0.897711
mesdames        0.902082
courage         1.024943
côte            1.062192
plage           1.419784
mademoiselle    1.421283
ors             1.747298
professeure     1.791751
enfance         2.015367
lendemains      2.045393
détail          2.122941
sables          2.232556
amitié          2.451087
paysanne        2.613066
monsieur        2.685090
vieille         2.819256
exemples        3.157083
madame          3.630800
Name: 355, dtype: float64

In [80]:
unique_freq_n_we['355'].sort_values()[:20]

poitrine     -17.556677
vêtement     -16.150639
seins        -16.125893
intérieur    -15.367145
suites       -15.325886
espèce       -14.876024
oreilles     -14.476932
ventre       -14.020979
pantalon     -13.736344
cuisses      -13.608708
ailes        -13.525702
intérieurs   -13.002431
blancs       -12.861234
poils        -12.774035
jambes       -12.376138
salles       -12.287629
fenêtre      -12.218022
rang         -12.040234
chemises     -12.032741
mecs         -11.648692
Name: 355, dtype: float64

## Dimensions with the lowest std

We can create a hypithesis that these dimensions with the lowest deviation, represent the noun-ness.

In [84]:
unique_freq_n_we.loc[:, :'511'].std().sort_values()[:5]

508    1.797948
408    1.843925
229    1.879214
497    1.898605
458    1.942430
dtype: float64

None of these dimensions were previously analysed.

# Testing the correlation of noun-ness

We now have 6 potential dimentions that hold the information about noun-ness: `181`, `508`, `408`, `229`, `497` and `458`.

We can test these dimensions for noun-ness by selecting top 1000 most frequent verbs (that can't be interpreted by nouns). For both nouns and verbs, we can create a column encoding the PoS: -1 for nouns and 1 for verbs. We can then check the correlation of each of the 6 dimensions with the PoS values.

First, we will create a DataFrame with words that are uniquely verbs:

In [89]:
top1000_unique_verbs = all_verbs_with_freq[~~all_verbs_with_freq.index.isin(all_nouns_with_freq.index)][:1000]
top1000_unique_verbs

lemme cgram  freqlemlivres
Word                                    
est            être   VER       15085.47
fût            être   VER       15085.47
sommes         être   VER       15085.47
étais          être   VER       15085.47
été            être   VER       15085.47
...             ...   ...            ...
accusés     accuser   VER          39.93
goûter       goûter   VER          39.86
renverse  renverser   VER          39.46
mouille    mouiller   VER          38.92
mouilles   mouiller   VER          38.92

[1000 rows x 3 columns]

In [90]:
all_v_we = pd.read_csv('all_verb_we.csv', index_col=0)
unique_freq_v_we = all_v_we.merge(top1000_unique_verbs, left_index=True, right_index=True)
unique_freq_v_we

0         1         2         3         4         5  \
a         1.621909 -1.563334 -2.426639 -1.438065 -1.388713 -6.022451   
abat      7.341515 -1.911545 -3.165352  0.881855  1.255615 -2.573494   
abats     2.571913 -0.997794  2.213261  7.866620  5.011807 -1.280126   
abattu    4.032405 -2.332013  3.859827 -0.144230 -1.695582 -2.872752   
abattus   7.670354 -0.896221  4.722162  0.242280  1.191507 -1.961131   
...            ...       ...       ...       ...       ...       ...   
étale     8.967206 -3.334843 -2.595041  2.622343  2.960083 -1.572292   
étendue   6.004179 -2.148462  8.073779  2.874971  3.621442 -5.073373   
étendues  3.499134 -0.164913  3.679914  2.277138  1.078756 -8.603376   
été       3.416336 -0.773586  1.706467 -0.731092 -2.286620 -4.172236   
être      0.166951 -1.121742  3.855680  0.846314 -3.065406  0.577783   

                 6         7         8         9  ...       505       506  \
a         0.529605  5.798126 -1.673958  2.689945  ... -2.933637 -3.008870   
abat     -0.540471 -2.959341  2.970952  1.989335  ... -1.498207 -4.692162   
abats    -1.463985 -5.969270 -0.341901 -3.891025  ...  2.976765 -6.461811   
abattu    1.532816  2.077215  4.141663  1.540514  ... -2.940739 -3.258436   
abattus   5.570742  0.197430 -2.174044  0.243627  ...  0.536509 -4.334290   
...            ...       ...       ...       ...  ...       ...       ...   
étale    -5.258777 -4.000706  2.279800  1.521071  ... -3.860589  1.245261   
étendue  -1.616805  4.065632  2.629171  0.531212  ... -0.564053  1.643511   
étendues -4.933097  4.446160  1.523909  1.621144  ... -0.481554 -1.865223   
été       0.249760  2.382638 -0.532814 -1.754949  ... -3.045682  3.401185   
être      1.696411  0.962430  1.131322 -5.577701  ... -2.119208  1.919185   

               507       508       509       510       511    lemme  cgram  \
a        -6.436573 -3.332313 -5.462238 -0.065580 -2.128896    avoir    VER   
abat     -2.878892 -6.178217 -0.410775  0.161333  3.765966  abattre    VER   
abats    -0.574586 -5.538427  2.004641  1.403862 -8.578699  abattre    VER   
abattu   -4.851328 -6.090186 -3.795757 -2.432890 -3.881667  abattre    VER   
abattus  -1.890132 -5.915509 -3.567384 -1.877470 -3.753840  abattre    VER   
...            ...       ...       ...       ...       ...      ...    ...   
étale    -3.233653 -2.511359 -2.267789 -1.174504  2.593932   étaler    VER   
étendue  -4.844989 -2.997791 -6.470501 -2.037538 -0.944526  étendre    VER   
étendues -3.761984 -0.568648 -5.724548 -2.363234  0.054831  étendre    VER   
été      -7.156822 -2.926211 -7.060627  4.555849 -4.830540     être    VER   
être     -4.168536 -1.392099 -5.050615  1.669238 -4.337863     être    VER   

          freqlemlivres  
a               6426.49  
abat              50.61  
abats             50.61  
abattu            50.61  
abattus           50.61  
...                 ...  
étale             63.11  
étendue          111.89  
étendues         111.89  
été            15085.47  
être           15085.47  

[819 rows x 515 columns]

Now we can merge the DataFrame of words that are uniquely nouns and uniquely verbs and construct a new dimension of PoS encoding.

In [119]:
unique_n_and_v_we = pd.concat([unique_freq_n_we, unique_freq_v_we])
unique_n_and_v_we

0         1         2         3         4         5  \
absence   3.322297 -0.768149  3.555035 -0.685460  1.201102 -3.151831   
absences  3.295355 -0.794079  2.472845  3.223941  5.035848 -4.848855   
accord   -0.732188 -2.024965  1.674970 -4.194049 -2.528312 -2.400746   
accords   1.037610 -0.246123  3.333316 -4.094496  0.056898 -2.978131   
acte      5.714411 -3.455246  1.156246  0.894613 -2.072708  1.619847   
...            ...       ...       ...       ...       ...       ...   
étale     8.967206 -3.334843 -2.595041  2.622343  2.960083 -1.572292   
étendue   6.004179 -2.148462  8.073779  2.874971  3.621442 -5.073373   
étendues  3.499134 -0.164913  3.679914  2.277138  1.078756 -8.603376   
été       3.416336 -0.773586  1.706467 -0.731092 -2.286620 -4.172236   
être      0.166951 -1.121742  3.855680  0.846314 -3.065406  0.577783   

                 6         7         8         9  ...       507       508  \
absence  -3.592423  1.189369 -2.742928 -2.666739  ... -3.189504 -1.781391   
absences -4.557629 -2.719433 -1.932022 -6.440993  ... -2.720232 -5.453354   
accord   -4.672920  0.374903  6.015941 -0.066068  ... -1.511964 -3.012614   
accords  -4.198946 -2.159259  3.082677  2.101236  ...  1.128479 -2.874813   
acte     -2.930497 -1.164060  1.456195 -3.298590  ... -3.608866 -1.903922   
...            ...       ...       ...       ...  ...       ...       ...   
étale    -5.258777 -4.000706  2.279800  1.521071  ... -3.233653 -2.511359   
étendue  -1.616805  4.065632  2.629171  0.531212  ... -4.844989 -2.997791   
étendues -4.933097  4.446160  1.523909  1.621144  ... -3.761984 -0.568648   
été       0.249760  2.382638 -0.532814 -1.754949  ... -7.156822 -2.926211   
être      1.696411  0.962430  1.131322 -5.577701  ... -4.168536 -1.392099   

               509       510       511        gn     gender    lemme  cgram  \
absence  -1.852207  5.123568 -2.381779  singular   feminine  absence    NOM   
absences -1.545623  5.919082  0.577700    plural   feminine  absence    NOM   
accord   -2.688764 -5.424189 -6.660925  singular  masculine   accord    NOM   
accords  -3.825052 -5.254343 -6.959417    plural  masculine   accord    NOM   
acte     -1.031769  1.625798 -3.859723  singular  masculine     acte    NOM   
...            ...       ...       ...       ...        ...      ...    ...   
étale    -2.267789 -1.174504  2.593932       NaN        NaN   étaler    VER   
étendue  -6.470501 -2.037538 -0.944526       NaN        NaN  étendre    VER   
étendues -5.724548 -2.363234  0.054831       NaN        NaN  étendre    VER   
été      -7.060627  4.555849 -4.830540       NaN        NaN     être    VER   
être     -5.050615  1.669238 -4.337863       NaN        NaN     être    VER   

          freqlemlivres  
absence           76.28  
absences          76.28  
accord           136.15  
accords          136.15  
acte              57.30  
...                 ...  
étale             63.11  
étendue          111.89  
étendues         111.89  
été            15085.47  
être           15085.47  

[1704 rows x 517 columns]

In [120]:
unique_n_and_v_we['pos'] = len(np.zeros(1704))

In [121]:
unique_n_and_v_we.loc[unique_n_and_v_we['cgram'] == 'NOM', 'pos'] = -1
unique_n_and_v_we.loc[unique_n_and_v_we['cgram'] == 'VER', 'pos'] = 1
unique_n_and_v_we

0         1         2         3         4         5  \
absence   3.322297 -0.768149  3.555035 -0.685460  1.201102 -3.151831   
absences  3.295355 -0.794079  2.472845  3.223941  5.035848 -4.848855   
accord   -0.732188 -2.024965  1.674970 -4.194049 -2.528312 -2.400746   
accords   1.037610 -0.246123  3.333316 -4.094496  0.056898 -2.978131   
acte      5.714411 -3.455246  1.156246  0.894613 -2.072708  1.619847   
...            ...       ...       ...       ...       ...       ...   
étale     8.967206 -3.334843 -2.595041  2.622343  2.960083 -1.572292   
étendue   6.004179 -2.148462  8.073779  2.874971  3.621442 -5.073373   
étendues  3.499134 -0.164913  3.679914  2.277138  1.078756 -8.603376   
été       3.416336 -0.773586  1.706467 -0.731092 -2.286620 -4.172236   
être      0.166951 -1.121742  3.855680  0.846314 -3.065406  0.577783   

                 6         7         8         9  ...       508       509  \
absence  -3.592423  1.189369 -2.742928 -2.666739  ... -1.781391 -1.852207   
absences -4.557629 -2.719433 -1.932022 -6.440993  ... -5.453354 -1.545623   
accord   -4.672920  0.374903  6.015941 -0.066068  ... -3.012614 -2.688764   
accords  -4.198946 -2.159259  3.082677  2.101236  ... -2.874813 -3.825052   
acte     -2.930497 -1.164060  1.456195 -3.298590  ... -1.903922 -1.031769   
...            ...       ...       ...       ...  ...       ...       ...   
étale    -5.258777 -4.000706  2.279800  1.521071  ... -2.511359 -2.267789   
étendue  -1.616805  4.065632  2.629171  0.531212  ... -2.997791 -6.470501   
étendues -4.933097  4.446160  1.523909  1.621144  ... -0.568648 -5.724548   
été       0.249760  2.382638 -0.532814 -1.754949  ... -2.926211 -7.060627   
être      1.696411  0.962430  1.131322 -5.577701  ... -1.392099 -5.050615   

               510       511        gn     gender    lemme  cgram  \
absence   5.123568 -2.381779  singular   feminine  absence    NOM   
absences  5.919082  0.577700    plural   feminine  absence    NOM   
accord   -5.424189 -6.660925  singular  masculine   accord    NOM   
accords  -5.254343 -6.959417    plural  masculine   accord    NOM   
acte      1.625798 -3.859723  singular  masculine     acte    NOM   
...            ...       ...       ...        ...      ...    ...   
étale    -1.174504  2.593932       NaN        NaN   étaler    VER   
étendue  -2.037538 -0.944526       NaN        NaN  étendre    VER   
étendues -2.363234  0.054831       NaN        NaN  étendre    VER   
été       4.555849 -4.830540       NaN        NaN     être    VER   
être      1.669238 -4.337863       NaN        NaN     être    VER   

          freqlemlivres  pos  
absence           76.28   -1  
absences          76.28   -1  
accord           136.15   -1  
accords          136.15   -1  
acte              57.30   -1  
...                 ...  ...  
étale             63.11    1  
étendue          111.89    1  
étendues         111.89    1  
été            15085.47    1  
être           15085.47    1  

[1704 rows x 518 columns]

In [122]:
unique_n_and_v_we.loc[:, [ '181', '508', '408', '229', '497', '458', 'pos']].corr()

181       508       408       229       497       458       pos
181  1.000000  0.031377 -0.007063 -0.010227  0.053273 -0.238095  0.208263
508  0.031377  1.000000  0.030125 -0.031889  0.039364  0.064826 -0.212633
408 -0.007063  0.030125  1.000000  0.134825  0.005695 -0.107360  0.007996
229 -0.010227 -0.031889  0.134825  1.000000  0.008892 -0.022842 -0.039496
497  0.053273  0.039364  0.005695  0.008892  1.000000  0.178985 -0.037110
458 -0.238095  0.064826 -0.107360 -0.022842  0.178985  1.000000 -0.279417
pos  0.208263 -0.212633  0.007996 -0.039496 -0.037110 -0.279417  1.000000

As we can see from the analysis, the strong correlation between the dimensions and `pos` is in the dimension `458` and there is no strong correlation between the dimensions themselves.

We can have a look at the dimensions with not only the most frequent nouns and verbs, but on the set of all words. 

In [126]:
unique_verbs = all_verbs_with_freq[~~all_verbs_with_freq.index.isin(all_nouns_with_freq.index)]
unique_verbs_we = all_v_we.merge(unique_verbs, left_index=True, right_index=True)
unique_nouns = all_nouns_with_freq[~~all_nouns_with_freq.index.isin(all_verbs_with_freq.index)]
unique_nouns_we = all_n_we.merge(unique_nouns, left_index=True, right_index=True)
all_unique = pd.concat([unique_verbs_we, unique_nouns_we])
all_unique.loc[all_unique['cgram'] == 'NOM', 'pos'] = -1
all_unique.loc[all_unique['cgram'] == 'VER', 'pos'] = 1
all_unique

0         1         2         3         4         5         6  \
a        1.621909 -1.563334 -2.426639 -1.438065 -1.388713 -6.022451  0.529605   
abat     7.341515 -1.911545 -3.165352  0.881855  1.255615 -2.573494 -0.540471   
abats    2.571913 -0.997794  2.213261  7.866620  5.011807 -1.280126 -1.463985   
abattu   4.032405 -2.332013  3.859827 -0.144230 -1.695582 -2.872752  1.532816   
abattus  7.670354 -0.896221  4.722162  0.242280  1.191507 -1.961131  5.570742   
...           ...       ...       ...       ...       ...       ...       ...   
été      3.416338 -0.773585  1.706465 -0.731093 -2.286620 -4.172235  0.249760   
évacué   6.187242 -6.105330  3.755269 -3.585482  2.970526 -5.341292  0.025965   
évacuée  6.950320 -5.794044  2.062860  0.921969 -1.625069 -5.335832 -1.289939   
évacués  3.478242 -3.813868  2.747407  0.856850  0.044580 -6.091687  1.160988   
être     0.166950 -1.121742  3.855680  0.846314 -3.065409  0.577782  1.696410   

                7         8         9  ...       508       509       510  \
a        5.798126 -1.673958  2.689945  ... -3.332313 -5.462238 -0.065580   
abat    -2.959341  2.970952  1.989335  ... -6.178217 -0.410775  0.161333   
abats   -5.969270 -0.341901 -3.891025  ... -5.538427  2.004641  1.403862   
abattu   2.077215  4.141663  1.540514  ... -6.090186 -3.795757 -2.432890   
abattus  0.197430 -2.174044  0.243627  ... -5.915509 -3.567384 -1.877470   
...           ...       ...       ...  ...       ...       ...       ...   
été      2.382637 -0.532811 -1.754951  ... -2.926212 -7.060628  4.555850   
évacué   2.611712  2.453145  1.364960  ... -6.745635 -2.194109  3.137106   
évacuée  4.485277  0.809356  3.228939  ... -7.385629 -2.920214  5.824216   
évacués  4.125816 -0.813236  1.399028  ... -7.141437 -3.194010  1.390625   
être     0.962429  1.131322 -5.577701  ... -1.392100 -5.050613  1.669239   

              511    lemme  cgram  freqlemlivres        gn     gender  pos  
a       -2.128896    avoir    VER        6426.49       NaN        NaN  1.0  
abat     3.765966  abattre    VER          50.61       NaN        NaN  1.0  
abats   -8.578699  abattre    VER          50.61       NaN        NaN  1.0  
abattu  -3.881667  abattre    VER          50.61       NaN        NaN  1.0  
abattus -3.753840  abattre    VER          50.61       NaN        NaN  1.0  
...           ...      ...    ...            ...       ...        ...  ...  
été     -4.830541      été    NOM         125.47  singular  masculine -1.0  
évacué  -6.339313   évacué    NOM           0.54  singular  masculine -1.0  
évacuée -0.765619   évacué    NOM           0.54  singular   feminine -1.0  
évacués -1.701988   évacué    NOM           0.54    plural  masculine -1.0  
être    -4.337866     être    NOM         122.57  singular  masculine -1.0  

[4238 rows x 518 columns]

In [127]:
all_unique.loc[:, [ '181', '508', '408', '229', '497', '458', 'pos']].corr()

181       508       408       229       497       458       pos
181  1.000000  0.061329 -0.030845  0.053523  0.027050 -0.191055 -0.005197
508  0.061329  1.000000 -0.012407  0.017735 -0.032003  0.001713 -0.003089
408 -0.030845 -0.012407  1.000000  0.092553 -0.039611 -0.064198 -0.002566
229  0.053523  0.017735  0.092553  1.000000 -0.025118 -0.017484  0.004509
497  0.027050 -0.032003 -0.039611 -0.025118  1.000000  0.185801  0.001342
458 -0.191055  0.001713 -0.064198 -0.017484  0.185801  1.000000 -0.001409
pos -0.005197 -0.003089 -0.002566  0.004509  0.001342 -0.001409  1.000000

We can see that with selecting more verbs and nouns the correlation between PoS becomes weaker.

# Finding dimensions with highest correlation to PoS values

Now we can attempt to find a dimension with the highest correlation to PoS values: 1 for verbs and -1 for nouns.

In [128]:
pos_corr_df = pd.DataFrame(columns=['corr_with_pos'])
pos_list = []

# Go through dimensions from 0 to 511
for i in range(512):
    # Find the correlation of the current column with "freqlemlivres" and add to the list
    pos_list.append(all_unique.loc[:, [str(i), 'pos']].corr().iloc[0][-1])

# Use the list values as 
pos_corr_df['corr_with_pos'] = freq_list
pos_corr_df['sign'] = np.sign(pos_corr_df['corr_with_pos'])
pos_corr_df['abs'] = abs(pos_corr_df['corr_with_pos'])
pos_corr_df

corr_with_pos  sign       abs
0        -0.056975  -1.0  0.056975
1         0.054485   1.0  0.054485
2        -0.014856  -1.0  0.014856
3        -0.053013  -1.0  0.053013
4        -0.020633  -1.0  0.020633
..             ...   ...       ...
507      -0.072089  -1.0  0.072089
508       0.013247   1.0  0.013247
509      -0.018811  -1.0  0.018811
510       0.049783   1.0  0.049783
511       0.026257   1.0  0.026257

[512 rows x 3 columns]

In [129]:
pos_corr_df.sort_values('abs')

corr_with_pos  sign       abs
439       0.000016   1.0  0.000016
378      -0.000033  -1.0  0.000033
240       0.000056   1.0  0.000056
141       0.000396   1.0  0.000396
67       -0.000428  -1.0  0.000428
..             ...   ...       ...
143      -0.110892  -1.0  0.110892
314       0.125546   1.0  0.125546
324      -0.129692  -1.0  0.129692
224      -0.131444  -1.0  0.131444
146      -0.133998  -1.0  0.133998

[512 rows x 3 columns]